### PR TITLE
Reduced diagnostics: fix calculation of particle energy in single precision 

### PR DESCRIPTION
**Pull request by lucafedeli88**
This PR rearranges the calculation of particle energy in reduced diags to avoid numerical errors when single precision is used. 
There were two issues:
- m^2 is too small to be represented with a `float`
- performing `sqrt(p^2c^2  + m^2c^4)-mc^2` in single precision gives 0 if momenta are small (as often happens for ions)  

Following @ax3l 's suggestion, I have introduced a gamma threshold (1.005) to switch between the full relativistic expression of the kinetic energy and its Taylor expansion (up to 10th order in |momentum| ) . I've also updated the reduced diag test accordingly, so that it performs the same calculation of the code.

❓ **Question:** should we still calculate gamma in double precision?

### CODE REVIEW

**Review by NeilZaim — APPROVED** [review]

**Comment by lucafedeli88** [comment]
Uhm, I was actually thinking to perform part of the calculation in double precision. When ux, uy, uz are very small, 1 + us in the square root becomes exactly equal to 1...

**Comment by NeilZaim** [comment]
Another option could be to revert to the classical formula (p^2/2m) if `us` beyond a given threshold, which may even be more accurate than switching to double precision. I think we had already discussed this issue at some point, but I don't remember exactly where.

**Comment by NeilZaim** [comment]
This is the discussion I was thinking about : https://github.com/ECP-WarpX/WarpX/pull/1908#discussion_r644508676

**Comment by lucafedeli88** [comment]
I see... it could be an interesting solution. Without choosing a threshold we may simply do something like:
```
auto ss = sqrt(1 + u^2) - 1 ;
auto res = ( ss > 0.0 ) ?  (ss):(u^2*0.5) ; 
```
i.e., if the kinetic energy is zero, try the classical calculation

**Comment by ax3l** [comment]
Maybe relevant: When I implemented this a while back in PIConGPU, I switched below a threshold for energy calcs to the non-relativistic formula to the low energy tail of ion spectras right:
- KinEnergy: https://github.com/ComputationalRadiationPhysics/picongpu/blob/0.5.0/include/picongpu/algorithms/KinEnergy.hpp

For Gamma calculations, we also did explicit double calcs where needed:
- Gamma: https://github.com/ComputationalRadiationPhysics/picongpu/blob/0.5.0/include/picongpu/algorithms/Gamma.hpp

Thresholds definitely give us a little bump when binning results exactly at the threshold value bins, because of the discontinuity.

**Comment by lucafedeli88** [comment]
> Maybe relevant: When I implemented this a while back in PIConGPU, I switched below a threshold for energy calcs to the non-relativistic formula to the low energy tail of ion spectras right:
> 
>     * KinEnergy: https://github.com/ComputationalRadiationPhysics/picongpu/blob/0.5.0/include/picongpu/algorithms/KinEnergy.hpp
> 
> 
> For Gamma calculations, we also did explicit double calcs where needed:
> 
>     * Gamma: https://github.com/ComputationalRadiationPhysics/picongpu/blob/0.5.0/include/picongpu/algorithms/Gamma.hpp
> 
> 
> Thresholds definitely give us a little bump when binning results exactly at the threshold value bins, because of the discontinuity.

Thanks! I've followed your suggestion about using a threshold. Do you think that we need to compute gamma in double precision even if we use a threshold?

**Comment by ax3l** [comment]
For the _energy_ itself, using the non-relativistic formula and all single precision was good enough in the past.

For the comparison `gamma > 1.005` we don't need a double precision calc, that's well in the significant digits of a `float`.
But if we use `gamma` at other places where we need a high precision if its low particle-energy values, then we should cast optionally.

File: Source/Diagnostics/ReducedDiags/ParticleEnergy.cpp [hunk 1]
```diff
@@ -159,8 +158,15 @@ void ParticleEnergy::ComputeDiags (int step)
...
                     const amrex::Real uz = p.rdata(PIdx::uz);
-                    const amrex::Real us = ux*ux + uy*uy + uz*uz;
-                    return {w*(std::sqrt(us*m2*c2 + m2*c4) - m*c2), w};
+                    const amrex::Real u2 = (ux*ux + uy*uy + uz*uz)*inv_c2;
+                    const auto gamma = std::sqrt(1.0_rt + u2);
+
+                    const auto kk = (gamma > PhysConst::gamma_relativistic_threshold)?
+                        (gamma-1.0_rt):
+                        (u2*0.5_rt - u2*u2*(1.0_rt/8.0_rt) + u2*u2*u2*(1.0_rt/16.0_rt)-
+                            u2*u2*u2*u2*(5.0_rt/128.0_rt) + (7.0_rt/256_rt)*u2*u2*u2*u2*u2); //Taylor expansion
```
Comment (ax3l) [inline Source/Diagnostics/ReducedDiags/ParticleEnergy.cpp:167]: Hihi, good idea to Taylor expand this as well.

Do you like to make this a functor/utility as in PIConGPU? I think we might want to reuse the logic in different places later on.

Comment (ax3l) [reply Source/Diagnostics/ReducedDiags/ParticleEnergy.cpp:167]: @lucafedeli88 I merge this for now, but can you please refactor this part out in a follow-up PR? :)

**Review by ax3l — APPROVED** [review]

File: Source/Diagnostics/ReducedDiags/ParticleEnergy.cpp (hunk 1 above)
Comment (lucafedeli88) [reply Source/Diagnostics/ReducedDiags/ParticleEnergy.cpp:167]: sure @ax3l ! where are we supposed to put these low-level functors? Should we create a "Physics" sub-folder in the Source folder?

**Pull request merged**